components/compositing/compositor.rs

```diff
@@ -1156,26 +1160,49 @@ impl<Window: WindowMethods> IOCompositor<Window> {
         // Batch up all scroll events into one, or else we'll do way too much painting.
         let mut last_combined_event: Option<ScrollZoomEvent> = None;
         for scroll_event in self.pending_scroll_zoom_events.drain(..) {
-            let this_delta = scroll_event.delta;
             let this_cursor = scroll_event.cursor;
+
+            let this_delta = match scroll_event.scroll_location {
+                ScrollLocation::Delta(delta) => delta,
+                _ => {
+                    // If this is an event which is scrolling to the start or end of the page,
+                    // disregard other pending events and exit the loop.
+                    last_combined_event = Some(scroll_event);
+                    break;
+                }
+            };
+
             if let Some(combined_event) = last_combined_event {
                 if combined_event.phase != scroll_event.phase {
-                    let delta = (combined_event.delta / self.scale).to_untyped();
-                    let cursor =
-                        (combined_event.cursor.to_f32() / self.scale).to_untyped();
-                    let delta = webrender_traits::LayerPoint::from_untyped(&delta);
-                    let location = webrender_traits::ScrollLocation::Delta(delta);
-                    let cursor = webrender_traits::WorldPoint::from_untyped(&cursor);
-                    self.webrender_api.scroll(location, cursor, combined_event.phase);
-                    last_combined_event = None
+                    if combined_event.phase != scroll_event.phase {
+                        let combined_delta = match combined_event.scroll_location {
```

Comment:
Is this code necessary? Won't the for loop have been broken out of by the condition above?